viamonvifdiscovery/service.go

```diff
@@ -0,0 +1,138 @@
+// Package viamonvifdiscovery provides the discovery service that wraps ONVIF integration for the viamrtsp module
+package viamonvifdiscovery
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/viam-modules/viamrtsp"
+	"github.com/viam-modules/viamrtsp/viamonvif"
+	"github.com/viam-modules/viamrtsp/viamonvif/device"
+	"go.viam.com/rdk/components/camera"
+	"go.viam.com/rdk/logging"
+	"go.viam.com/rdk/resource"
+	"go.viam.com/rdk/services/discovery"
+)
+
+// Model is the model for a rtsp discovery service.
+var (
+	Model             = viamrtsp.Family.WithModel("onvif")
+	errNoCamerasFound = errors.New("no cameras found, ensure cameras are working or check credentials")
+)
+
+func init() {
+	resource.RegisterService(
+		discovery.API,
+		Model,
+		resource.Registration[discovery.Service, *Config]{
+			Constructor: newDiscovery,
+		})
+}
+
+// Config is the config for the discovery service.
+type Config struct {
+	Credentials []device.Credentials `json:"credentials"`
+}
+
+// Validate validates the discovery service.
+func (cfg *Config) Validate(_ string) ([]string, error) {
+	// check that all creds have both usernames and passwords set. Note a credential can have both fields empty
+	for _, cred := range cfg.Credentials {
+		if cred.Pass != "" && cred.User == "" {
+			return nil, fmt.Errorf("credential missing username, has password %v", cred.Pass)
+		}
+	}
+	return []string{}, nil
+}
+
+type rtspDiscovery struct {
+	resource.Named
+	resource.AlwaysRebuild
+	resource.TriviallyCloseable
+	Credentials []device.Credentials
+	logger      logging.Logger
+}
+
+func newDiscovery(_ context.Context, _ resource.Dependencies,
+	conf resource.Config,
+	logger logging.Logger,
+) (discovery.Service, error) {
+	cfg, err := resource.NativeConfig[*Config](conf)
+	if err != nil {
+		return nil, err
+	}
+	dis := &rtspDiscovery{
+		Named: conf.ResourceName().AsNamed(),
+		// place the empty credentials first, so we find the insecure cameras before other cameras.
```

Comment:
what's empty about the credentials here? Aren't we just pulling in the deserialized value from the config

Reply:
oh this is another out of date comment that I missed